### ISSUE TITLE
How to get the actual framebuffer content?

### ISSUE THREAD

**Issue by ichlubna** [aap/librw#138]
Hi @aap. Thanks for this library! I use it for some scientific experiments and I am struggling getting the content of a framebuffer. Would you please explain to me how your pipeline works? I have tried off-screen rendering where I bound a new framebuffer instead of the default one at all places where `glDraw...` is used. Having issues with that, I tried to go to the very end of the drawing and did this simple thing in `glDevice.cpp` in `showRaster` function:

```
std::vector<uint8> data(raster->width*raster->height*4);
glReadBuffer(GL_BACK);
glReadPixels(0,0,raster->width,raster->height,GL_BGRA,GL_UNSIGNED_BYTE,data.data());
glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, image->width, image->height, 0, GL_RGBA, GL_UNSIGNED_BYTE, image->pixels);
```
I tried to render the texture but it does not contain the same content as I would see in the game. Please would you kindly tell me how to obtain the image as it is viewed when the game plays? 

I am trying to render the scene in several views, similarly to VR mods. My plan is to add a loop somewhere around the `glDraw` commands, edit the matrices and render only to a portion of the framebuffer. Or maybe an easier solution would be to force the engine to render everything several times and call the `showRaster` only once at the end while storing the framebuffer in each iteration (would be again related to the issue above)? Then I need to do something with the views as described above. Any suggestions regarding my approach are welcomed.
Thank you!

**Comment by aap** [comment]
scientific experiments? woah. anyway, if you want to get the pixel values into main memory for something like a screenshot i think what should work is camera->frameBuffer->toImage(). locking that raster should read out the pixels similar to how you tried to do it. but i'm not entirely sure i understand what you want to do, reading the pixel values to main memory may not be the right solution. perhaps the subrast example is more like what you need?

**Comment by ichlubna** [comment]
Thank you for the quick reply @aap. Imagine my problem as producing a VR-like side-by-side views and some post-process. So what I need is:

1. Render the scene several times with different camera settings
2. Get the views in one shader and process them

I tried to start with the second point by trying to get the framebuffer to a texture and render in a new draw call on a fullscreen quad. This is where I am stuck right now :D. So in fact, I don't need it in the main memory, just desperately tried the simplest solution. The optimal solution for me would be to render everything into different parts of an off-screen framebuffer and then project it on a fullscreen quad. However, I see that you do some tricks with the framebuffers so simple binding to my own framebuffer before calling glDraw... does not seem to be enough.

**Comment by ichlubna** [comment]
> perhaps the subrast example is more like what you need?

Would you be so kind @aap and explain a little but the way subrast works? I tried to edit the librw code with the goal to render two views and display them. I am really confused about the main camera and subcameras in your example. This is just a test I tried in `camera.cpp` with the goal to add a new little red window in the resulting camera:

```
Camera *subCam;
rw::RGBA BackgroundColor = { 255, 0, 0, 255 };

void
defaultBeginUpdateCB(Camera *cam)
{
   engine->currentCamera = cam;
    Frame::syncDirty();
    // cloned the same camera as the main one
    subCam = cam->clone();
    // prepared the place for the view
    rw::Rect rect;
    rect.x = 0;
    rect.y = 0;
    rect.w = 50;
    rect.h = 50;
    // tried to inject the view into the parent
    subCam->frameBuffer->subRaster(cam->frameBuffer, &rect);
    // made the view red
    subCam->clear(&BackgroundColor, rw::Camera::CLEARIMAGE|rw::Camera::CLEARZ);
    engine->device.beginUpdate(subCam);
    engine->device.beginUpdate(cam);
}

void
defaultEndUpdateCB(Camera *cam)
{
       engine->device.endUpdate(cam);
       engine->device.endUpdate(subCam);
}
```
I am trying to inject my code into librw where any input camera would produce multiple views without the need to create more cameras on the higher level.

**Comment by aap** [comment]
I have no idea what you're doing there. first of all you're creating a new camera every frame without ever getting rid of it? that sounds bad. second: you can only render to one camera at a time. calling beginUpdate twice on different cameras is not how this works at all. I think what you want is to render the scene twice? I'm not entirely sure what your goal is.

**Comment by ichlubna** [comment]
> I think what you want is to render the scene twice? 

Yes, sorry for the lack of explanation. I'd ideally love to take the camera the user uses with librw, render it, translate it a bit, and render again, showing both views side by side. So that nothing needs to be changed in the project, only changing the code of librw.
And I know that camera creation every frame etc. was bad, just wanted to test the idea of cloning the camera in librw and mixing the views. I don't know how to reproduce the same thing as in subrast example but inside librw itself.

**Comment by ichlubna** [comment]
> I have no idea what you're doing there. 
Here is a visual example. I can adjust the camera position by:
```
V3d translation{1.0f, 0.0f, 0.0f};
cam->getFrame()->translate(&translation);
```
which is the left image but I struggle with the simultaneous rendering and showing at the same time as here:

![Image](https://github.com/user-attachments/assets/15811144-163c-4577-9692-7daec5b5b242)

**Comment by aap** [comment]
Doing this at the librw layer sounds like a bad idea

**Comment by ichlubna** [comment]
> Doing this at the librw layer sounds like a bad idea

OK, thanks for the verdict and sorry for bothering you. So to make it work, if I render the scene with one camera, clone it, translate, render again, I should have two views, each stored in the camera->frameBuffer, right?

**Comment by aap** [comment]
I would not clone the camera more than once. normally the camera's framebuffer is the screen. you can either render to different regions of the screen with subrasters, or you render to a camera texture and compose the image from your two textures in the end.

**Comment by ichlubna** [comment]
Thanks to your advice @aap I am now able to render to different parts of the main raster but for some reason, I am not able to make each camera differently positioned. I implemented the thing on the re3 level. I iterate over the cameras, render to the main camera raster but when I try to transform the Frame, it happens only once for the first camera. Do you have any idea why this happens? I have also read the RenderWare docs but I can't find the reason. I do the loop [here](https://github.com/ichlubna/re3LKG/blob/miami/src/core/main.cpp#L1566) and transform the camera [later](https://github.com/ichlubna/re3LKG/blob/miami/src/core/main.cpp#L1608). It seems like I can't even render to the other cameras properly if the first one is not used which is weird. Sorry for the repeated subquestions :D.

**Comment by aap** [comment]
all i know is that you're trying to use RW/librw in a really weird way. i couldn't even tell you what i'd expect your code to do in the first place tbh.

**Comment by ichlubna** [comment]
I guess I miss some fundamentals about librw. According to your advice, the docs and the subrast example I tried to do the following. Simplified and updated code:

1. Created an array of new cameras in a loop where I made every camera's raster a subraster of the main scene one.
```
camera[i] = CameraCreate(camWidth, camHeight, TRUE));
rw::Rect rect{col*camWidth, row*camHeight, camWidth, camHeight};
RwCameraGetRaster(camera[i])->subRaster(RwCameraGetRaster(Scene.camera), &rect);
RpWorldAddCamera(Scene.world, camera[i]);
```

2. Found where re3 renders the scene. There, added a loop there that temporarily replaces the main scene camera with the new ones. The rendering happens in the loop for each camera. Then I return back the original scene camera to make the game show its raster at the end.

```
// They do the rendering here
void
Idle(void *arg)
{
// Backup the original camera
auto originalCamera = Scene.camera;
// Looping over my cameras created above
for(int i=0; i<holoCameraCount; i++)
{
    // Temporarily replacing the main scene camera so that all rendering functions below render to my new camera
    Scene.camera = holoCameras[i];
    // Creating new transform matrix to avoid using the same frame for all cameras
    rw::Matrix *transform = rw::Matrix::create();
    // Making the transform the same as the scene camera
    transform->setIdentity();
    rw::Matrix *originalTransform = originalCamera->getFrame()->getLTM();
    rw::Matrix::mult(transform, transform, originalTransform);
    // Moving the camera so that each is positioned differently
    rw::V3d trans{i*1.0f,0.0f,0.0f};
    transform->translate(&trans);
    Scene.camera->getFrame()->transform(transform, rw::COMBINEREPLACE);
    ..... the re3 rendering stuff (left unchanged) like:
   CTimer::Update();
   CGame::Process();
   RenderScene(); 
   etc....
}
// Returning the original camera
Scene.camera = originalCamera;
// Rest of the function unchanged
if (gbShowTimebars)
	tbDisplay();
```

The subraster part seems to work well but the transformation does not change at all. It seems like the rendering depends on the first camera transformation and is locked afterwards. If I add something like:
```
if(i == 0)
  trans = {1.0f,0.0f,0.0f};
```
The cameras move but all the same no matter what transform is applied in other loop iterations. This is the result:

![Image](https://github.com/user-attachments/assets/6a78511e-b0e4-46aa-bd25-a7415fb34cf8)

**Comment by ichlubna** [comment]
A little update @aap , I just found out that what locks the transform is probably this:
```
TheCamera.RenderMotionBlur();
```
Probably the post processing does something that I miss. When I comment the function out, the transformation starts working. Not sure what exactly happens there though.

![Image](https://github.com/user-attachments/assets/3dacdc85-6824-4ba8-8783-5feeff74bf4f)
